src/browser_utils.js

```diff
@@ -1081,6 +1085,132 @@ async function getSelectOptions(page, select) {
     }, select);
 }
 
+/**
+ *
+ * @param {import('./config').Config} config
+ * @param {import('puppeteer').Page} page
+ * @param {string} fileName
+ * @param {string} [selector]
+ */
+async function takeScreenshot(config, page, fileName, selector) {
+    await mkdirp(config.screenshot_directory);
+    const fn = path.join(config.screenshot_directory, fileName);
+
+    const viewport = page.viewport();
+    let img;
+    if (selector) {
+        const el = await page.waitForSelector(selector);
+        img = await el.screenshot({
+            path: fn,
+            type: 'png',
+        });
+    } else {
+        img = await page.screenshot({
+            path: fn,
+            type: 'png',
+            fullPage: true,
+        });
+    }
+
+    // Restore emulation, fixes unable to resize window after taking a screenshot.
+    await page._client.send('Emulation.clearDeviceMetricsOverride');
+
+    // Restore potential emulation settings that were active before
+    // we took the screenshot.
+    if (viewport !== null) {
+        await page.setViewport(viewport);
+    }
+
+    return img;
+}
+
+/**
+ * @typedef {"minor" | "moderate" | "serious" | "critical"} A11yImpact
+ */
+
+/**
+ * @typedef {{html: string, screenshots: Array<Buffer | null>, selectors: string[]}} A11yNode
+ */
+
+/**
+ *
+ * @typedef {{impact: A11yImpact, helpUrl?: string, description: string, nodes: A11yNode[]}} A11yResult
+ */
+
+/**
+ *
+ * @param {import('./config').Config}
+ * @param {import('puppeteer').Frame | import('puppeteer').Page} page
+ */
+async function assertAccessibility(config, page) {
+    assert(config, 'Missing config argument');
+    assert(page, 'Missing page argument');
+
+    output.logVerbose(config, '[a11y] Checking for accessibility errors...');
+
+    const url = page.url();
+
+    await page.addScriptTag({
+        path: require.resolve('axe-core')
+    });
+
+    /** @type {import('axe-core').AxeResults} */
+    const results = await page.evaluate(() => {
+        return new Promise((resolve, reject) => {
+            window.axe.run(document, { ancestry: true }, (err, results) => {
+                if (err !== null) reject(err);
+                else resolve(results);
+            });
+        });
+    });
+
+    const errors = config.accessibilityErrors;
+
+    let i = errors.length;
+    for (const v of results.violations) {
+
+        /** @type {A11yNode[]} */
+        const nodes = [];
+
+        for (const node of v.nodes) {
+            let imgs = [];
+            if (node.ancestry) {
+                // We can't postpone taking screenshots as the html may change later
+                const name = `${config._taskName}-a11y-${i++}`;
+                imgs = await Promise.all(node.ancestry.map(async selector => {
```

Comment:
What's going on here – aren't we taking multiple screenshots to the same file in parallel?